protocol-units/da-sequencer/node/src/celestia/mod.rs

```diff
@@ -1,101 +1,150 @@
+	/// Send the given block to Celestia.
+	/// The block is not immediately sent but aggregated in a blob
+	/// until the client can send it to celestia.
+	fn send_block(
 		&self,
-		block: &SequencerBlockDigest,
-	) -> std::result::Result<(), DaSequencerError> {
+		_block: SequencerBlockDigest,
+	) -> Pin<Box<dyn Future<Output = Result<(), DaSequencerError>> + Send + '_>> {
 		todo!()
 	}
 
 	/// Get the blob from celestia at the given height.
-	pub async fn get_blob_at_height(&self) -> Blob {
+	fn get_blobs_at_height(
+		&self,
+		_height: CelestiaHeight,
+	) -> Pin<Box<dyn Future<Output = Result<Option<Vec<Blob>>, DaSequencerError>> + Send + '_>> {
 		todo!()
 	}
 
-	/// Bootstrap the Celestia client to recovert from missing block.
+	/// Bootstrap the Celestia client to recover from missing block.
 	/// In case of crash for example, block sent to Celestia can be behind the block created by the network.
-	/// The role of this function is to recovert this missing block to all block of the network are sent to celestia.
-	/// The basic algorythm is start from 'last_sent_block_height' to 'current_block_height' and request using the notifier channel
-	/// The missing block. Not sure last_notified_celestia_height is useful.
+	/// The role of this function is to recover this missing block to all block of the network are sent to celestia.
+	/// The basic algorithm is start from 'last_sent_block_height' to 'current_block_height' and request using the notifier channel
+	/// the missing block. Not sure last_notified_celestia_height is useful.
 	/// During this boostrap new block are sent to the client.
 	/// These block should be buffered until the boostrap is done then sent after in order.
-	pub async fn bootstrap(
+	fn bootstrap(
 		&self,
 		current_block_height: BlockHeight,
-		last_sent_block_height: BlockHeight,
-		last_notified_celestia_height: CelestiaHeight,
-	) -> std::result::Result<(), DaSequencerError> {
-		todo!()
+		_last_sent_block_height: BlockHeight,
+		last_finalized_celestia_height: CelestiaHeight,
+	) -> Pin<Box<dyn Future<Output = Result<(), DaSequencerError>> + Send + '_>> {
+		Box::pin(async move {
+			// wait to ensure that no blob is pending in the Celestia network
+			tokio::time::sleep(DELAY_SECONDS_BEFORE_BOOTSTRAPPING).await;
+
+			// Step 1: Get last digest in the last finalized blob
+			let blobs = self.get_blobs_at_height(last_finalized_celestia_height).await?;
+			// TODO: Handle all of this gracefully with a DaSequencerError
+			let blobs = blobs.expect("get_blobs_at_height returned None");
+			let last = blobs.last().expect("get_blobs_at_height returned no Blobs");
+			let _digest = last.0.last().expect("get_blobs_at_height returned an empty Blob");
+
+			// Step 2: Get the Block for digest
+			// TODO: Request the block for digest
+			let block: SequencerBlock = SequencerBlock::default();
+
+			// Step 3: Request and send all missing blocks
+			for height in block.get_height().0..=current_block_height.0 {
```

Comment:
You should skip the `block.get_height()` because it has already been sent.

Reply:
Right.